### PR TITLE
Validate the hash salt during object creation

### LINKED ISSUE #233: Hasher.validateSalt(salt) is underutilized
**Describe the bug**
The various impls of Hasher (e.g. HashCodeHasher and ModHashCodeHasher) have constructors that take salt parameters that aren't validated.  So you can do something like:

`ModHashCodeHasher hasher = new ModHashCodeHasher(200, "42");
`

but you can't subsequently call 

`hasher.setSalt("42");`

because that calls 

`Hasher.validateSalt("42");
`

which returns false. 

Seems like this is inconsistent and should be remedied by having constructors that take a salt value call the validateSalt method.

**System information:**

 - Tribuo Version: 4.2.x

### PR DESCRIPTION
**Pull request by geoffreydstewart**
### Description
These changes add validations of the hash salt during object creation, either programmatically or via the Configuration System.

### Motivation
These changes fix #233

### CODE REVIEW

File: Core/src/main/java/org/tribuo/hash/ModHashCodeHasher.java [hunk 1]
```diff
@@ -73,7 +74,13 @@ public ModHashCodeHasher(int dimension, String salt) {
      * Used by the OLCUT configuration system, and should not be called by external code.
      */
     @Override
-    public void postConfig() {
+    public synchronized void postConfig() throws PropertyException {
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/hash/ModHashCodeHasher.java:77]: No need for synchronized.

File: Core/src/main/java/org/tribuo/hash/ModHashCodeHasher.java [hunk 2]
```diff
@@ -73,7 +74,13 @@ public ModHashCodeHasher(int dimension, String salt) {
...
+        if (salt == null) {
+            throw new PropertyException("","salt","Salt not set in ModHashCodeHasher.");
+        }
+        else if (!Hasher.validateSalt(salt)) {
```
Comment (Craigacp) [inline Core/src/main/java/org/tribuo/hash/ModHashCodeHasher.java:81]: Please move the else if onto the line above.

**Review by Craigacp — CHANGES_REQUESTED** [review]
Small formatting things mostly. The tests look good, thanks for adding them.

File: Core/src/main/java/org/tribuo/hash/HashCodeHasher.java [hunk 3]
```diff
@@ -59,6 +61,19 @@ public String hash(String name) {
         return ""+salted.hashCode();
     }
 
+    /**
+     * Used by the OLCUT configuration system, and should not be called by external code.
+     */
+    @Override
+    public synchronized void postConfig() throws PropertyException{
```
[the remark this answers is not in the document]
Comment (geoffreydstewart) [reply Core/src/main/java/org/tribuo/hash/HashCodeHasher.java]: Sure, it makes sense to only apply synchronization as needed to safely modify an attribute. I think I may have misunderstood that from a previous discussion. Thanks for clarifying.

[the remark this answers is not in the document]
Comment (Craigacp) [reply Core/src/main/java/org/tribuo/hash/HashCodeHasher.java]: Yeah, the false positive only arises in `Trainer`s as those also access the RNG field under synchronization. In practice as `postConfig` is always only called either from the constructor, or from OLCUT before the object is published it doesn't actually need synchronization, but the static analysis tools just see a public method and so warn on it.

**Pull request merged**